File: sdk/python/workersql_client/streaming.py
"""
Result streaming support for WorkerSQL Python SDK
Enables processing large result sets without loading all data into memory
"""

from typing import Iterator, List, Dict, Any, Optional, Callable, Generator
import time


class QueryStream:
    """Streaming query result reader with iterator protocol"""

    def __init__(
        self,
        sql: str,
        params: Optional[List[Any]] = None,
        query_fn: Callable[[str, Optional[List[Any]]], Any] = None,
        batch_size: int = 100
    ):
        self.sql = sql
        self.params = params or []
        self.query_fn = query_fn
        self.batch_size = batch_size
        self.offset = 0
        self.ended = False
        self.current_batch: List[Dict[str, Any]] = []
        self.batch_index = 0

    def __iter__(self) -> Iterator[Dict[str, Any]]:
        """Make this class iterable"""
        return self

    def __next__(self) -> Dict[str, Any]:
        """Get next row from stream"""
        # If we have rows in current batch, return next one
        if self.batch_index < len(self.current_batch):
            row = self.current_batch[self.batch_index]
            self.batch_index += 1
            return row

        # If we're done, stop iteration
        if self.ended:
            raise StopIteration

        # Fetch next batch
        self.current_batch = self._fetch_next_batch()
        self.batch_index = 0

        if not self.current_batch:
            self.ended = True
            raise StopIteration

        row = self.current_batch[self.batch_index]
        self.batch_index += 1
        return row

    def _fetch_next_batch(self) -> List[Dict[str, Any]]:
        """Fetch next batch of rows"""
        stream_sql = self._add_pagination(self.sql, self.batch_size, self.offset)
        result = self.query_fn(stream_sql, self.params)
        rows = result.data or []

        self.offset += len(rows)

        # If we got fewer rows than batch_size, we've reached the end
        if len(rows) < self.batch_size:
            self.ended = True

        return rows

    def _add_pagination(self, sql: str, limit: int, offset: int) -> str:
        """Add LIMIT and OFFSET to SQL query"""
        trimmed_sql = sql.strip()

        # Remove existing LIMIT clause if present
        import re
        limit_pattern = r'\s+LIMIT\s+\d+(\s+OFFSET\s+\d+)?$'
        base_sql = re.sub(limit_pattern, '', trimmed_sql, flags=re.IGNORECASE)

        return f'{base_sql} LIMIT {limit} OFFSET {offset}'

    def fetchmany(self, size: Optional[int] = None) -> List[Dict[str, Any]]:
        """Fetch multiple rows at once"""
        size = size or self.batch_size
        rows = []

        try:
            for _ in range(size):
                rows.append(next(self))
        except StopIteration:
            pass

        return rows

    def fetchall(self) -> List[Dict[str, Any]]:
        """Fetch all remaining rows"""
        rows = []

        try:
            while True:
                rows.append(next(self))
        except StopIteration:
            pass

        return rows

    def close(self):
        """Close the stream"""
        self.ended = True
        self.current_batch = []


class AsyncQueryIterator:
    """Async iterator for query results (for use with async/await)"""

    def __init__(
        self,
        sql: str,
        params: Optional[List[Any]] = None,
        query_fn: Callable = None,
        batch_size: int = 100
    ):
        self.sql = sql
        self.params = params or []
        self.query_fn = query_fn
        self.batch_size = batch_size
        self.offset = 0
        self.current_batch: List[Dict[str, Any]] = []
        self.batch_index = 0
        self.done = False

    def __aiter__(self):
        """Make this class async iterable"""
        return self

    async def __anext__(self) -> Dict[str, Any]:
        """Get next row asynchronously"""
        # If we have rows in current batch, return next one
        if self.batch_index < len(self.current_batch):
            row = self.current_batch[self.batch_index]
            self.batch_index += 1
            return row

        # If we're done, stop iteration
        if self.done:
            raise StopAsyncIteration

        # Fetch next batch
        await self._fetch_next_batch()

        if not self.current_batch:
            self.done = True
            raise StopAsyncIteration

        row = self.current_batch[self.batch_index]
        self.batch_index += 1
        return row

    async def _fetch_next_batch(self):
        """Fetch next batch of rows asynchronously"""
        stream_sql = self._add_pagination(self.sql, self.batch_size, self.offset)
        result = await self.query_fn(stream_sql, self.params)
        self.current_batch = result.data or []
        self.batch_index = 0
        self.offset += len(self.current_batch)

        if len(self.current_batch) < self.batch_size:
            self.done = True

    def _add_pagination(self, sql: str, limit: int, offset: int) -> str:
        """Add LIMIT and OFFSET to SQL query"""
        trimmed_sql = sql.strip()

        import re
        limit_pattern = r'\s+LIMIT\s+\d+(\s+OFFSET\s+\d+)?$'
        base_sql = re.sub(limit_pattern, '', trimmed_sql, flags=re.IGNORECASE)

        return f'{base_sql} LIMIT {limit} OFFSET {offset}'


class CursorStream:
    """Cursor-based streaming for large result sets"""

    def __init__(
        self,
        sql: str,
        params: Optional[List[Any]] = None,
        query_fn: Callable = None,
        batch_size: int = 100
    ):
        self.sql = sql
        self.params = params or []
        self.query_fn = query_fn
        self.batch_size = batch_size
        self.cursor_id: Optional[str] = None
        self.closed = False

    def open(self):
        """Open cursor and prepare for streaming"""
        try:
            # Initialize cursor (implementation depends on server support)
            result = self.query_fn(
                f'DECLARE cursor_{int(time.time())} CURSOR FOR {self.sql}',
                self.params
            )

            self.cursor_id = result.cursor_id if hasattr(result, 'cursor_id') else f'cursor_{int(time.time())}'
            return self.cursor_id
        except Exception as error:
            raise Exception(f'Failed to open cursor: {error}')

    def fetch_next(self) -> List[Dict[str, Any]]:
        """Fetch next batch of rows"""
        if self.closed or not self.cursor_id:
            return []

        try:
            result = self.query_fn(
                f'FETCH {self.batch_size} FROM {self.cursor_id}',
                []
            )

            rows = result.data or []

            if not rows:
                self.close()

            return rows
        except Exception as error:
            self.close()
            raise Exception(f'Failed to fetch from cursor: {error}')

    def close(self):
        """Close cursor and release resources"""
        if self.closed or not self.cursor_id:
            return

        try:
            self.query_fn(f'CLOSE {self.cursor_id}', [])
            self.closed = True
        except Exception:
            # Ignore close errors
            pass

    def __enter__(self):
        """Context manager support"""
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager cleanup"""
        self.close()
        return False


def create_query_stream(
    sql: str,
    params: Optional[List[Any]],
    query_fn: Callable,
    batch_size: int = 100
) -> QueryStream:
    """Helper function to create a query stream"""
    return QueryStream(sql, params, query_fn, batch_size)


def create_async_iterator(
    sql: str,
    params: Optional[List[Any]],
    query_fn: Callable,
    batch_size: int = 100
) -> AsyncQueryIterator:
    """Helper function to create an async iterator"""
    return AsyncQueryIterator(sql, params, query_fn, batch_size)

File: sdk/node/src/streaming.ts
/**
 * Result streaming support for WorkerSQL Node.js SDK
 * Enables processing large result sets without loading all data into memory
 */

import { Readable } from 'stream';
import { EventEmitter } from 'events';

export interface StreamOptions {
  highWaterMark?: number;
  batchSize?: number;
  timeout?: number;
}

export interface StreamRow {
  [key: string]: any;
}

/**
 * Streaming query result reader
 */
export class QueryStream extends Readable {
  private offset: number = 0;
  private ended: boolean = false;

  constructor(
    private sql: string,
    private params: any[] = [],
    private queryFn: (sql: string, params?: any[]) => Promise<any>,
    private options: StreamOptions = {}
  ) {
    super({ objectMode: true, highWaterMark: options.highWaterMark || 16 });
    this.options.batchSize = options.batchSize || 100;
  }

  async _read(): Promise<void> {
    if (this.ended) {
      this.push(null);
      return;
    }

    try {
      // Modify SQL to add LIMIT and OFFSET
      const streamSql = this.addPagination(this.sql, this.options.batchSize!, this.offset);

      const result = await this.queryFn(streamSql, this.params);
      const rows = result.data || [];

      if (rows.length === 0) {
        this.ended = true;
        this.push(null);
        return;
      }

      for (const row of rows) {
        if (!this.push(row)) {
          // Backpressure - stop reading
          return;
        }
      }

      this.offset += rows.length;

      // If we got fewer rows than batchSize, we've reached the end
      if (rows.length < this.options.batchSize!) {
        this.ended = true;
        this.push(null);
      }
    } catch (error) {
      this.destroy(error as Error);
    }
  }

  private addPagination(sql: string, limit: number, offset: number): string {
    // Simple pagination - in production, would need more sophisticated SQL parsing
    const trimmedSql = sql.trim();

    // Remove existing LIMIT clause if present
    const limitPattern = /\s+LIMIT\s+\d+(\s+OFFSET\s+\d+)?$/i;
    const baseSql = trimmedSql.replace(limitPattern, '');

    return `${baseSql} LIMIT ${limit} OFFSET ${offset}`;
  }
}

/**
 * Cursor-based streaming for large result sets
 */
export class CursorStream extends EventEmitter {
  private cursorId: string | null = null;
  private closed: boolean = false;

  constructor(
    private sql: string,
    private params: any[] = [],
    private queryFn: (sql: string, params?: any[]) => Promise<any>,
    private options: StreamOptions = {}
  ) {
    super();
    this.options.batchSize = options.batchSize || 100;
  }

  /**
   * Open cursor and start streaming
   */
  async start(): Promise<void> {
    try {
      // Initialize cursor (implementation depends on server support)
      const result = await this.queryFn(
        `DECLARE cursor_${Date.now()} CURSOR FOR ${this.sql}`,
        this.params
      );

      this.cursorId = result.cursorId || `cursor_${Date.now()}`;
      this.emit('open', this.cursorId);

      await this.fetchNext();
    } catch (error) {
      this.emit('error', error);
    }
  }

  /**
   * Fetch next batch of rows
   */
  private async fetchNext(): Promise<void> {
    if (this.closed || !this.cursorId) {
      return;
    }

    try {
      const result = await this.queryFn(
        `FETCH ${this.options.batchSize} FROM ${this.cursorId}`,
        []
      );

      const rows = result.data || [];

      if (rows.length === 0) {
        await this.close();
        this.emit('end');
        return;
      }

      for (const row of rows) {
        this.emit('data', row);
      }

      // Continue fetching
      setImmediate(() => this.fetchNext());
    } catch (error) {
      this.emit('error', error);
      await this.close();
    }
  }

  /**
   * Close cursor and release resources
   */
  async close(): Promise<void> {
    if (this.closed || !this.cursorId) {
      return;
    }

    try {
      await this.queryFn(`CLOSE ${this.cursorId}`, []);
      this.closed = true;
      this.emit('close');
    } catch (error) {
      this.emit('error', error);
    }
  }
}

/**
 * Async iterator for query results
 */
export class QueryIterator implements AsyncIterableIterator<StreamRow> {
  private offset: number = 0;
  private currentBatch: StreamRow[] = [];
  private batchIndex: number = 0;
  private done: boolean = false;

  constructor(
    private sql: string,
    private params: any[] = [],
    private queryFn: (sql: string, params?: any[]) => Promise<any>,
    private batchSize: number = 100
  ) {}

  async next(): Promise<IteratorResult<StreamRow>> {
    // If we have rows in current batch, return next one
    if (this.batchIndex < this.currentBatch.length) {
      return {
        done: false,
        value: this.currentBatch[this.batchIndex++]
      };
    }

    // If we're done, return
    if (this.done) {
      return { done: true, value: undefined };
    }

    // Fetch next batch
    try {
      const streamSql = this.addPagination(this.sql, this.batchSize, this.offset);
      const result = await this.queryFn(streamSql, this.params);
      this.currentBatch = result.data || [];
      this.batchIndex = 0;
      this.offset += this.currentBatch.length;

      if (this.currentBatch.length === 0) {
        this.done = true;
        return { done: true, value: undefined };
      }

      if (this.currentBatch.length < this.batchSize) {
        this.done = true;
      }

      return {
        done: false,
        value: this.currentBatch[this.batchIndex++]
      };
    } catch (error) {
      this.done = true;
      throw error;
    }
  }

  [Symbol.asyncIterator](): AsyncIterableIterator<StreamRow> {
    return this;
  }

  private addPagination(sql: string, limit: number, offset: number): string {
    const trimmedSql = sql.trim();
    const limitPattern = /\s+LIMIT\s+\d+(\s+OFFSET\s+\d+)?$/i;
    const baseSql = trimmedSql.replace(limitPattern, '');
    return `${baseSql} LIMIT ${limit} OFFSET ${offset}`;
  }
}

/**
 * Helper functions for creating streaming queries
 */
export function createQueryStream(
  sql: string,
  params: any[],
  queryFn: (sql: string, params?: any[]) => Promise<any>,
  options?: StreamOptions
): QueryStream {
  return new QueryStream(sql, params, queryFn, options);
}

export function createQueryIterator(
  sql: string,
  params: any[],
  queryFn: (sql: string, params?: any[]) => Promise<any>,
  batchSize?: number
): QueryIterator {
  return new QueryIterator(sql, params, queryFn, batchSize);
}
